modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/RemoveConstantsFromQueryTest.java

```diff
@@ -0,0 +1,179 @@
+                "UPDATE PUBLIC.TST SET NAME = ? WHERE ID = ?"),
+
+            F.t("DELETE FROM TST WHERE name = 'Sarah Connor'",
+                "DELETE FROM PUBLIC.TST WHERE NAME = ?"),
+
+            F.t("SELECT * FROM TST WHERE name = 'Sarah Connor'",
+                "SELECT __Z0.ID, __Z0.NAME, __Z0.AGE FROM PUBLIC.TST __Z0 WHERE __Z0.NAME = ?"),
+
+            F.t("SELECT * FROM TST WHERE name = SUBSTR('Sarah Connor', 0, 2)",
+                "SELECT __Z0.ID, __Z0.NAME, __Z0.AGE FROM PUBLIC.TST __Z0 WHERE __Z0.NAME = ?"),
+
+            F.t("SELECT * FROM TST GROUP BY id HAVING name = 'X'",
+                "SELECT __Z0.ID, __Z0.NAME, __Z0.AGE FROM PUBLIC.TST __Z0 GROUP BY __Z0.ID HAVING __Z0.NAME = ?"),
+
+            F.t("SELECT * FROM TST GROUP BY id HAVING name = 'X' UNION SELECT * FROM TST GROUP BY id HAVING name = 'Y'",
+                "(SELECT __Z0.ID, __Z0.NAME, __Z0.AGE FROM PUBLIC.TST __Z0 GROUP BY __Z0.ID HAVING __Z0.NAME = ?)" +
+                    " UNION " +
+                    "(SELECT __Z1.ID, __Z1.NAME, __Z1.AGE FROM PUBLIC.TST __Z1 GROUP BY __Z1.ID HAVING __Z1.NAME = ?)"),
+
+            F.t("SELECT CONCAT(name, 'xxx') FROM TST",
+                "SELECT CONCAT(__Z0.NAME, ?) FROM PUBLIC.TST __Z0"),
+
+            F.t("ALTER TABLE TST ADD COLUMN department VARCHAR(200)", null),
+
+            F.t("DROP TABLE TST", null),
+
+            F.t("KILL SERVICE 'my_service'", null)
+        );
+
+        AtomicReference<String> lastQryFromEvt = new AtomicReference<>();
+
+        ignite.events().localListen(evt -> {
+            assertTrue(evt instanceof SqlQueryExecutionEvent);
+
+            lastQryFromEvt.set(((SqlQueryExecutionEvent)evt).text());
+
+            return true;
+        }, EventType.EVT_SQL_QUERY_EXECUTION);
+
+        for (IgniteBiTuple<String, String> qry : qries) {
+            execSql(ignite, qry.get1());
+
+            String expHist = qry.get2() == null ? qry.get1() : qry.get2();
+
+            assertEquals(expHist, lastQryFromEvt.get());
+
+            List<List<?>> hist = execSql(ignite, "SELECT sql FROM SYS.SQL_QUERIES_HISTORY WHERE sql = ?", expHist);
+
+            assertNotNull(hist);
+            assertEquals(1, hist.size());
+            assertEquals(hist.get(0).get(0), expHist);
+        }
+
+        Set<String> qriesFromStats = new HashSet<>();
+
+        stopCollectStatisticsAndRead(new AbstractPerformanceStatisticsTest.TestHandler() {
+            @Override public void query(
+                UUID nodeId,
+                GridCacheQueryType type,
+                String text,
+                long id,
+                long startTime,
+                long duration,
+                boolean success
+            ) {
+                qriesFromStats.add(text);
+            }
+        });
+
+        assertEquals(qries.size(), qriesFromStats.size());
+
+        assertTrue(qriesFromStats.contains("SELECT SQL FROM SYS.SQL_QUERIES_HISTORY WHERE SQL = ?1"));
```

Comment:
Why `?1`? I didn't get this syntax.

Reply:
H2 parser works like that.

Reply:
It's acknowledged by any contract? Perhaps it's some H2 bug and we rely on this again.

Reply:
Yes. Please, take a look at `GridSqlParameter#getSQL()`

Reply:
Ok, thanks